### PR TITLE
FISH-790 Refactor, fix bugs in  JavaEE Context Utility

### PR DESCRIPTION
**Pull request by lprimak**
## Description
Refactored Context utilities into immutable, serializable class.
Fixed many bugs along the way.

### Blockers
This PR is a blocker for Hazelcast 4 integration.

### New tests
Added a unit test

### Testing Performed
All test suites pass, tested with Hazelcast 4 branch

### Testing Environment
Java 8, Mac

## Documentation
Documentation added via Internal API javadoc for new EE context API

## Notes for Reviewers
This issue includes APIs necessary for Hazelcast 4 to be much more tightly integrated with Jakarta EE and Payara
when you think `invocation.preInvoke()` this is easy-to-use replacement and so much more!
This refactoring has been in the works for 6+ months, so it's not "new" in any sense of the word

### CODE REVIEW

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
Fixes #5000

File: api/payara-api/src/main/java/fish/payara/context/ContextProducer.java
Comment (pdudits) [inline api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: Is this really something we encourage users to use in their applications, and promise to keep API binary compatible?

I think this should have home elsewhere, somewhere under `nucleus/common`.

File: appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/CDIExtension.java
Comment (pdudits) [inline appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/CDIExtension.java:53]: Making it user-facing API is apparently your intent here, I'm not convinced that we should.

**Review by jbee — COMMENTED** [review]
I did a first rough review. I feel I would want way more description of what this PR tries to do, what problems are behind it. As far as I could see this mostly adds a new API and its implementation but isn't used in so many places. Where it is used the main change appears to be to use an instance (something got bound I guess) instead of what corresponds to a generic context. Before I look closer I really would like to get better understanding so please add some paragraphs to the description of the PR that help with that. Thanks.

File: api/payara-api/src/main/java/fish/payara/context/ContextProducer.java
Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: nucleus/common is where I started with this before migrating to Payara-api.
This was designed before MP context propagation, and perhaps we should use that in the future,
but it's not a reason to hold up this PR currently.

File: nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/JavaEEContextUtil.java
Comment (lprimak) [reply nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/JavaEEContextUtil.java:51]: The internal API is a superset of the external one now, which is correct.

File: appserver/web/weld-integration/src/main/java/org/glassfish/cdi/transaction/TransactionScopedCDIUtil.java
Comment (lprimak) [reply appserver/web/weld-integration/src/main/java/org/glassfish/cdi/transaction/TransactionScopedCDIUtil.java:222]: Context Util API changed, this line just tracks that change, it would be a compile error otherwise.
It's self-explanatory for me here and shouldn't require a comment, as it would be like
"tracking API changes"

File: appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/ContextualizerImpl.java
Comment (lprimak) [reply appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/ContextualizerImpl.java:55]: Javadoc is in the Contextualizer API itself

File: api/payara-api/src/main/java/fish/payara/context/ContextProducer.java
Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:108]: When I use this in NetBeans, it always puts in `ContextProducer.Interface` and I would not use Interface as a standalone type, so IMHO this isn't an issue

**Comment by lprimak** [comment]
@jbee Please read the Javadoc in the new public API. This should explain most, if not all of your questinos

File: api/payara-api/src/main/java/fish/payara/context/ContextProducer.java
Comment (pdudits) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: There's also Java EE Concurrency API, that deals with context propagation. But neither of these APIs expose internals of application servers, like invocation stack or underlying classloaders and give explicit control over them to user application. They only allow propagating current context to async calls.

I don't believe this belongs into application server's public api.

Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: Do you have an alternative? How else is Hazelcast callback supposed to deal with EJB/CDI objects?
I don't look at it as 'App server internals' rather than an integration point that let's apps use Jakarta EE.

Other frameworks always touted 'we have a main method' type approach, this gives and entry-way to do something like that on a hybrid basis at least, which is very useful

Comment (pdudits) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: I cannot offer an alternative, because I don't know what the exact requirements are. Does any of the other PRs demonstrate how would user app use these?

Jakarta EE is all about integration, yet so far they seemed to solve this without giving fine-grained control over invocation context to user. I can give you lots of examples that are probably not relevant, but may help understanding the motivation. And then we can look at whether the API could be more opaque, and then look at security implications of this.

Is there anything in JCA's inbound flows (so special kind of message driven bean) that would cover these use cases? Could CDI events be produced by application server instead of user app directly interacting with hazelcast over callback? An interceptor that would get installed around Hazelcast callback by the application server and manage the stack?

Comment (pdudits) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: Ah I now saw the example in the PR description, I can see few alternatives to write it:

Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: I edited the top of the PR with an example.
You do raise a good point about security, as this API makes it much easier to switch context into another application running on the same server (if ran without security manager)

Comment (pdudits) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: I found out, that I'm not entirely sure what the code sample is trying to achive.
```

        map.put("hello", "world"); // (1)

        map2.put("bye", listener); // (2)
```
What I think that (1) does is, that when when entry is added to map, it will invoke `@Inject MyListener listener` asynchronoulsy in current context. i.e invocation of Singleton EJB.

For this I can imagine
- Using `javax.enterprise.concurrent.ContextService`, its api is strinkingly similar.
- Set of Hazelcast specific APIs, i. e. CDI extension that would install map listeners and then invoke event observers (`void entryAdded(@Observes @EntryAdded(mapName = "hallo") String entry) {}`)


I'm not sure what (2) is trying to achieve. What should happen when we put managed object into distributed map? How do we manage it serializes properly since it the reference might be just a proxy (to a proxy) of the application class? How do we guarantee that map is updated when objects (or proxies) internal state changes? How do we cope with tha fact, that the grid is usually not running single version of an application and therefore serialization compatibility of the managed object must be retained? What if effects of invoking a method on the objects are local only?

Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: Jakarta EE has 'solved' - not really? this problem by mandating that objects can be used only in App server control flow. If we want true Hazelcast integration (which people are asking for), some sort of context API needs to be exposed.

Also, keep this in mind: https://github.com/payara/Payara/pull/5013
I am also working on tenant control for Hazelcast which will make it possible for some Hazelcast objects to automatically propagate Payara app context, but some cannot and still need an API to do this

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastSerializer.java
Comment (lprimak) [reply nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastSerializer.java:62]: fixed

File: api/payara-api/src/main/java/fish/payara/context/ContextProducer.java
Comment (lprimak) [reply api/payara-api/src/main/java/fish/payara/context/ContextProducer.java:60]: > Set of Hazelcast specific APIs, i. e. CDI extension that would install map listeners and then invoke event observers (void entryAdded(@Observes @EntryAdded(mapName = "hallo") String entry) {})

This is clearly out of scope for this PR

> Using javax.enterprise.concurrent.ContextService, its api is strinkingly similar.

Again, this was pre-MP context propagation, and if we are going to use something, it's MP stuff I would think
I thin it would also be out of scope here

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
@jbee @pdudits Since the public API is such a hot button issue and this PR is meant as a bug fix, I moved it into internal API so none of the above is an issue anymore

**Comment by lprimak** [comment]
jenkins test

**Comment by jbee** [comment]
@lprimak 

> Please read the Javadoc in the new public API. This should explain most, if not all of your questinos

Code tells me what is going on not why. I'd like to know you motivations both towards the problem and the chosen solution so I can understand this change in context.

**Review by jbee — APPROVED** [review]
I dont't see any obvious issues with the PR but likewise it is hard to tell how or where it would introduce new issues. This is just an area that is hard to verify other then understanding the consequences of the change and its impact which is a hard nut to crack for this one. I'm quite convinced this is an improvement over the old code but ultimately I feel this is your call @MattGill98 being responsible for community.

**Review by pdudits — APPROVED** [review]

**Pull request merged**
